### PR TITLE
CB-15730 - CM server moved during upgrade to another node

### PR DESCRIPTION
**Pull request by pdarvasi**
This commit introduces consistent hostname generation and consistent primary gateway selection logic across multiple recoveries, meaning:
- it is guaranteed that
    - the instanceId-s remain the same
    - the privateId allocation remains the same
    - primary gateways remain the same

### CODE REVIEW

**Review by gergopapi2 — APPROVED** [review]

File: cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsMetadataCollector.java [hunk 1]
```diff
@@ -81,8 +82,8 @@
                 ac.getCloudContext().getName(), resources.size(), vms.size(), allInstances.size());
         try {
             List<String> knownInstanceIdList = allInstances.stream()
-                    .filter(cloudInstance -> cloudInstance.getInstanceId() != null)
                     .map(CloudInstance::getInstanceId)
+                    .filter(Objects::nonNull)
```
[the remark this answers is not in the document]
Comment (pdarvasi) [reply cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsMetadataCollector.java:86]: It is not better, but

1. a method reference is more elegant than a lambda
2. IntelliJ Idea marked it as a potential optimisation

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java [hunk 2]
```diff
@@ -293,7 +293,10 @@ private void selectPrimaryGWFromGatewayInstances(Collection<InstanceGroup> insta
                 .findFirst();
 
         if (gwInstanceGroup.isPresent()) {
-            List<InstanceMetaData> gwInstances = instanceMetaDataService.findAllByInstanceGroupAndInstanceStatus(gwInstanceGroup.get(), InstanceStatus.CREATED);
+            // Ordered list of metadata to guarantee consistent primary gateway generation across multiple cluster recoveries
```
Comment (sodre90) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java:296]: If we had primary GW then we select it here: https://github.com/hortonworks/cloudbreak/blob/master/core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java#L185-L189
So in upgrade (repair) we restore the PGW to the node with the same FQDN. I don't understand why this change needed?
